_source/_posts/2025-07-08-express-authentication.md

```diff
@@ -0,0 +1,687 @@
+
+export const userTeamMap = {
+  "hannah.smith@task-vantage.com": "admin",
+  "grace.li@task-vantage.com": "legal",
+  "frank.wilson+@task-vantage.com": "dev advocacy",
+  "carol.lee@task-vantage.com": "finance",
+  "alice.johnson@task-vantage.com": "marketing",
+  "sarah.morgan@task-vantage.com": "hr",
+};
+
+export const dummyExpenseData = {
+  finance: [
+    {
+      name: "Alice Johnson",
+      item: "Product Launch Campaign",
+      amount: 1200,
+    },
+    {
+      name: "Bob Smith",
+      item: "Promotional Material",
+      amount: 450,
+    },
+    {
+      name: "Carol Lee",
+      item: "Team Lunch",
+      amount: 180,
+    },
+    {
+      name: "David Kim",
+      item: "Event Booth",
+      amount: 950,
+    },
+  ],
+  hr: [
+    {
+      name: "Eve Martinez",
+      item: "Internet",
+      amount: 300,
+    },
+    {
+      name: "Frank Wilson",
+      item: "Compliance Training",
+      amount: 600,
+    },
+    {
+      name: "Grace Li",
+      item: "Conference Travel",
+      amount: 1500,
+    },
+    {
+      name: "Henry Zhang",
+      item: "Team Offsite",
+      amount: 1000,
+    },
+  ],
+  marketing: [
+    {
+      name: "Alice Johnson",
+      item: "Payroll Processing",
+      amount: 750,
+    },
+    {
+      name: "Carol Lee",
+      item: "Compliance Training",
+      amount: 400,
+    },
+    {
+      name: "Eve Martinez",
+      item: "Team Lunch",
+      amount: 200,
+    },
+    {
+      name: "Frank Wilson",
+      item: "Team Offsite",
+      amount: 850,
+    },
+  ],
+  legal: [
+    {
+      name: "Grace Li",
+      item: "Event Booth",
+      amount: 1100,
+    },
+    {
+      name: "David Kim",
+      item: "Product Launch Campaign",
+      amount: 1300,
+    },
+    {
+      name: "Bob Smith",
+      item: "Conference Travel",
+      amount: 1250,
+    },
+    {
+      name: "Henry Zhang",
+      item: "Team Lunch",
+      amount: 170,
+    },
+  ],
+  "dev-advocacy": [
+    {
+      name: "Eve Martinez",
+      item: "Internet",
+      amount: 280,
+    },
+    {
+      name: "Frank Wilson",
+      item: "Payroll Processing",
+      amount: 720,
+    },
+    {
+      name: "Grace Li",
+      item: "Compliance Training",
+      amount: 500,
+    },
+    {
+      name: "Alice Johnson",
+      item: "Team Offsite",
+      amount: 950,
+    },
+  ],
+};
+```
+
+### Create an auth file to handle authentication
+
+The authentication logic is kept in a separate file, `auth.js`, to keep your server organized and maintainable. This file manages OpenID Connect authentication with [openid-client](https://www.passportjs.org/packages/openid-client), including PKCE support. It sets up the OIDC client, handles login and logout, processes callbacks, and provides middleware to protect routes.
+
+
+```javascript
+import * as client from "openid-client";
+import "dotenv/config";
+
+import { userTeamMap, ALL_TEAMS_NAME } from './utils.js';
+
+export function ensureAuthenticated(req, res, next) {
```

Comment:
We need to discuss what's going on here. This should be the main focus of the blog. What is the `ensureAuthticated` function, how does PKCE work in this library, what are you doing with the callbackURLParams function, etc etc